ggml-org/llama.cpp
# Add support for DeepseekV2ForCausalLM

This pull request adds support for DeepseekV2ForCausalLM-based models. Both lite and non-lite models are supported. Fixes #7118

Changes included in this pull request:

- increase max number of experts to 160
- add new LLM_ARCH_DEEPSEEK2 architecture
- add new model header parameters:
  - leading_dense_block_count - number of leading dense layers (following layers are MoE)
  - expert_feed_forward_length - ff length for experts (feed_forward_length is used for dense ffn dimension in leading dense layers)
  - expert_shared_count - number of shared experts
  - expert_weights_scale - in DeepSeekV2 MoE gate_inp output is not normalized, but multiplied by this constant value instead
  - attention.q_lora_rank - latent compressed Q dimension
  - attention.kv_lora_rank - latent compressed KV dimension
  - rope.scaling.yarn_log_multiplier - log multiplier used to calculate YaRN mscale (0.0707 in DeepSeek-V2 instead of 0.1)
- add new tensors (first hidden representations are mapped to latent compressed representation with _a tensors and normalized with _a_norm, then mapped to q/kv space with _b tensors):
  - attn_q_a
  - attn_q_b
  - attn_kv_a_mqa
  - attn_kv_b
  - attn_q_a_norm
  - attn_kv_a_norm
- add MODEL_16B and MODEL_236B model types for lite and non-lite models (would be great to have dense + expert param counts, but I have no idea how to count it properly considering the existence of leading dense layer, shared experts etc.)
- add support for DeepseekV2ForCausalLM model conversion in convert-hf
- add support for inference for models based on LLM_ARCH_DEEPSEEK2

This pull request also fixes #7331 by removing the failing assertion.

Note that I had to change the llm_build_moe_ffn() API to add scaling of MoE gate output - added a bool indicating whether to scale or not and a float scale value. Let me know if there is a better way.
Implementation also required somewhat ugly workaround for problems with YaRN implementation that I described in detail in #7416

One thing I'm not sure about - shall I print all new parameter values in llm_load_print_meta() or print them only for LLM_ARCH_DEEPSEEK2 models (since they are specific to this architecture)? Or perhaps simply not print them at all?

I'll correct the whitespace formatting after we agree on the parameter names.
